File: flutter/packages/aval_format/lib/src/manifest_validation.dart
/// Shared manifest schema validation primitives.
///
/// Dart port of `packages/format/src/manifest-validation.ts`. Manifest input
/// values are represented as plain `Object?` / `Map<String, Object?>` /
/// `List<Object?>`, mirroring how the TS source treats `unknown` JSON input
/// before it is validated into the typed `model.dart` classes.
library;

import 'constants.dart' show identifierPattern, sha256HexPattern;
import 'errors.dart';
import 'model.dart' show ResidencyEndpointV01;
import 'utf8.dart' show utf8ByteLength;

Map<String, Object?> record(Object? value, String path) {
  if (value is! Map) {
    invalid(path, 'must be an object');
  }
  return value.map((key, v) => MapEntry(key as String, v));
}

List<Object?> array(Object? value, String path) {
  final result = _arrayValue(value, path);
  _requireDenseArray(result, path);
  return result;
}

List<Object?> _arrayValue(Object? value, String path) {
  if (value is! List) {
    invalid(path, 'must be an array');
  }
  return value;
}

void _requireDenseArray(List<Object?> value, String path) {
  // Dart Lists are always dense; this mirrors the TS sparse-array check as a
  // no-op guard kept for structural parity with the source.
}

List<Object?> boundedArray(
  Object? value,
  String path,
  int minimum,
  int maximum,
) {
  final result = _arrayValue(value, path);
  if (result.length < minimum || result.length > maximum) {
    invalid(path, 'must contain between $minimum and $maximum entries');
  }
  _requireDenseArray(result, path);
  return result;
}

List<Object?> tuple(Object? value, int length, String path) {
  final result = _arrayValue(value, path);
  if (result.length != length) {
    invalid(path, 'must contain exactly $length entries');
  }
  _requireDenseArray(result, path);
  return result;
}

void exactKeys(
  Map<String, Object?> value,
  List<String> required, [
  String path = '',
  List<String> optional = const [],
]) {
  final allowed = {...required, ...optional};
  for (final key in value.keys) {
    if (!allowed.contains(key)) {
      invalid(path, 'contains unknown field ${quote(key)}');
    }
  }
  for (final key in required) {
    if (!owns(value, key)) {
      invalid('$path.$key', 'is required');
    }
  }
}

bool owns(Map<String, Object?> value, String key) => value.containsKey(key);

String generatorString(Object? value, String path) {
  if (value is! String) {
    invalid(path, 'must be a string');
  }
  for (var index = 0; index < value.length; index += 1) {
    if (value.codeUnitAt(index) <= 0x1f) {
      invalid(path, 'must not contain C0 controls');
    }
  }
  final length = utf8ByteLength(value, (message, [offset]) {
    invalid(path, 'contains a lone surrogate');
  });
  if (length < 1 || length > 128) {
    invalid(path, 'must contain between 1 and 128 UTF-8 bytes');
  }
  return value;
}

String identifier(Object? value, String path) {
  if (value is! String || !identifierPattern.hasMatch(value)) {
    invalid(path, 'must match ${identifierPattern.pattern}');
  }
  return value;
}

String digest(Object? value, String path) {
  if (value is! String || !sha256HexPattern.hasMatch(value)) {
    invalid(path, 'must be a lowercase 64-character SHA-256 hexadecimal string');
  }
  return value;
}

int positiveInteger(Object? value, String path, [int maximum = _maxSafeInteger]) {
  return integerInRange(value, path, 1, maximum);
}

int nonNegativeInteger(Object? value, String path) {
  return integerInRange(value, path, 0, _maxSafeInteger);
}

const int _maxSafeInteger = 9007199254740991;

int integerInRange(Object? value, String path, int minimum, int maximum) {
  if (value is! int || value < minimum || value > maximum) {
    invalid(path, 'must be a safe integer from $minimum to $maximum');
  }
  return value;
}

T literal<T>(Object? value, T expected, String path) {
  if (value != expected) {
    invalid(path, 'must be ${quote(expected.toString())}');
  }
  return expected;
}

String oneOf(Object? value, List<String> choices, String path) {
  if (value is! String || !choices.contains(value)) {
    invalid(path, 'must be one of ${choices.map(quote).join(', ')}');
  }
  return value;
}

/// Generic over any type with a string identifier, matching the TS
/// structural type `readonly { readonly id: string }[]`.
void requireIdOrder<T>(List<T> values, String Function(T) idOf, String path) {
  requireStringOrder(values.map(idOf).toList(), path);
}

void requireStringOrder(List<String> values, String path) {
  for (var index = 1; index < values.length; index += 1) {
    if (compareAscii(values[index - 1], values[index]) >= 0) {
      invalid(path, 'must be sorted by ID and contain no duplicates');
    }
  }
}

void requireNumberOrder(List<int> values, String path) {
  for (var index = 1; index < values.length; index += 1) {
    if (values[index - 1] >= values[index]) {
      invalid(path, 'must be numerically sorted and unique');
    }
  }
}

int compareEndpoint(ResidencyEndpointV01 a, ResidencyEndpointV01 b) {
  final byState = compareAscii(a.state, b.state);
  return byState != 0 ? byState : compareAscii(a.port, b.port);
}

int compareAscii(String a, String b) => a == b ? 0 : (a.compareTo(b) < 0 ? -1 : 1);

String quote(String value) => '"$value"';

Never invalid(String path, String message) {
  throw FormatError(
    FormatErrorCode.manifestInvalid,
    '$path $message',
    FormatErrorDetails(path: path),
  );
}

File: packages/format/src/manifest-validation.ts
import { IDENTIFIER_PATTERN, SHA256_HEX_PATTERN } from "./constants.js";
import { FormatError } from "./errors.js";
import { utf8ByteLength } from "./utf8.js";
import type { ResidencyEndpoint } from "./model.js";

export function record(value: unknown, path: string): Record<string, unknown> {
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    invalid(path, "must be an object");
  }
  return value as Record<string, unknown>;
}

export function array(value: unknown, path: string): readonly unknown[] {
  const result = arrayValue(value, path);
  requireDenseArray(result, path);
  return result;
}

function arrayValue(value: unknown, path: string): readonly unknown[] {
  if (!Array.isArray(value)) {
    invalid(path, "must be an array");
  }
  return value;
}

function requireDenseArray(value: readonly unknown[], path: string): void {
  for (let index = 0; index < value.length; index += 1) {
    if (!owns(value, String(index))) {
      invalid(`${path}[${String(index)}]`, "must not be sparse");
    }
  }
}

export function boundedArray(
  value: unknown,
  path: string,
  minimum: number,
  maximum: number
): readonly unknown[] {
  const result = arrayValue(value, path);
  if (result.length < minimum || result.length > maximum) {
    invalid(
      path,
      `must contain between ${String(minimum)} and ${String(maximum)} entries`
    );
  }
  requireDenseArray(result, path);
  return result;
}

export function tuple(
  value: unknown,
  length: number,
  path: string
): readonly unknown[] {
  const result = arrayValue(value, path);
  if (result.length !== length) {
    invalid(path, `must contain exactly ${String(length)} entries`);
  }
  requireDenseArray(result, path);
  return result;
}

export function exactKeys(
  value: Record<string, unknown>,
  required: readonly string[],
  path: string,
  optional: readonly string[] = []
): void {
  const allowed = new Set([...required, ...optional]);
  for (const key of Reflect.ownKeys(value)) {
    if (typeof key !== "string" || !allowed.has(key)) {
      invalid(
        path,
        `contains unknown field ${typeof key === "string" ? quote(key) : "[symbol]"}`
      );
    }
  }
  for (const key of required) {
    if (!owns(value, key)) {
      invalid(`${path}.${key}`, "is required");
    }
  }
}

export function owns(value: object, key: string): boolean {
  return Object.prototype.hasOwnProperty.call(value, key);
}

export function generatorString(value: unknown, path: string): string {
  if (typeof value !== "string") {
    invalid(path, "must be a string");
  }
  for (let index = 0; index < value.length; index += 1) {
    if (value.charCodeAt(index) <= 0x1f) {
      invalid(path, "must not contain C0 controls");
    }
  }
  const length = utf8ByteLength(value, () =>
    invalid(path, "contains a lone surrogate")
  );
  if (length < 1 || length > 128) {
    invalid(path, "must contain between 1 and 128 UTF-8 bytes");
  }
  return value;
}

export function identifier(value: unknown, path: string): string {
  if (typeof value !== "string" || !IDENTIFIER_PATTERN.test(value)) {
    invalid(path, `must match ${String(IDENTIFIER_PATTERN)}`);
  }
  return value;
}

export function digest(value: unknown, path: string): string {
  if (typeof value !== "string" || !SHA256_HEX_PATTERN.test(value)) {
    invalid(path, "must be a lowercase 64-character SHA-256 hexadecimal string");
  }
  return value;
}

export function positiveInteger(
  value: unknown,
  path: string,
  maximum = Number.MAX_SAFE_INTEGER
): number {
  return integerInRange(value, path, 1, maximum);
}

export function nonNegativeInteger(value: unknown, path: string): number {
  return integerInRange(value, path, 0, Number.MAX_SAFE_INTEGER);
}

export function integerInRange(
  value: unknown,
  path: string,
  minimum: number,
  maximum: number
): number {
  if (
    typeof value !== "number" ||
    !Number.isSafeInteger(value) ||
    value < minimum ||
    value > maximum
  ) {
    invalid(
      path,
      `must be a safe integer from ${String(minimum)} to ${String(maximum)}`
    );
  }
  return value;
}

export function literal<T extends string | number>(
  value: unknown,
  expected: T,
  path: string
): T {
  if (value !== expected) {
    invalid(path, `must be ${quote(String(expected))}`);
  }
  return expected;
}

export function oneOf<const T extends readonly string[]>(
  value: unknown,
  choices: T,
  path: string
): T[number] {
  if (typeof value !== "string" || !choices.includes(value)) {
    invalid(path, `must be one of ${choices.map(quote).join(", ")}`);
  }
  return value as T[number];
}

export function requireIdOrder(
  values: readonly { readonly id: string }[],
  path: string
): void {
  requireStringOrder(
    values.map((value) => value.id),
    path
  );
}

export function requireStringOrder(values: readonly string[], path: string): void {
  for (let index = 1; index < values.length; index += 1) {
    if (compareAscii(values[index - 1]!, values[index]!) >= 0) {
      invalid(path, "must be sorted by ID and contain no duplicates");
    }
  }
}

export function requireNumberOrder(values: readonly number[], path: string): void {
  for (let index = 1; index < values.length; index += 1) {
    if (values[index - 1]! >= values[index]!) {
      invalid(path, "must be numerically sorted and unique");
    }
  }
}

export function compareEndpoint(
  a: ResidencyEndpoint,
  b: ResidencyEndpoint
): number {
  return compareAscii(a.state, b.state) || compareAscii(a.port, b.port);
}

export function compareAscii(a: string, b: string): number {
  return a < b ? -1 : a > b ? 1 : 0;
}

export function quote(value: string): string {
  return JSON.stringify(value);
}

export function invalid(path: string, message: string): never {
  throw new FormatError("MANIFEST_INVALID", `${path} ${message}`, { path });
}
